### PR TITLE
PCA based loss landscape plotting callback added

### PR DESCRIPTION
**Pull request by johnviljoen**
Hi - this is a callback which saves the training trajectory in parameter space, finds the first two principal components, and uses them along with a basis of the trained optimal vector of parameters to plot a loss landscape (see https://arxiv.org/abs/1712.09913).

Its intended usage is as follows:

    callback = PCA_Callback()
    
    trainer = nm.Trainer(
        problem,
        train_loader,
        dev_loader,
        dev_loader,
        optimizer=optimizer,
        callback=callback,
        epochs=epochs,
        train_metric="train_loss",
        dev_metric="dev_loss",
        eval_metric='dev_loss',
    )
    
    trainer.train()
    
    callback.plot_pca_trajectory(trainer=trainer, num_points=20)

### CODE REVIEW

**Comment by RBirmiwal** [comment]
Hi John, 

This is something that I think can be easily used with one of the LIghtning hooks, training_step() might suffice depending on when you want it to execute. I think one can pass in callbacks to a LightningModule via callbacks=[callback1, callback2, etc.]  If that's possible let's use Lightning wherever possible
